File: Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (!head || !head->next) return false;
        auto slow = head;
        auto fast = head->next;
        while (slow != fast) {
            slow = slow->next;
            fast = fast->next;

            if (!fast) return false;

            fast = fast->next;
            if (!fast) return false;
        }
        return true;
    }
};

File: Linked List Cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
    if head == nil || head.Next == nil {
        return false
    }

    slow := head
    fast := head.Next
    for slow != fast {
        slow = slow.Next
        fast = fast.Next

        if fast == nil {
            return false
        }

        fast = fast.Next
        if fast == nil {
            return false
        }
    }
    return true
}
